libsolidity/ast/Types.cpp

```diff
@@ -905,6 +916,11 @@ TypePointer RationalNumberType::binaryOperatorResult(Token::Value _operator, Typ
 			uint32_t exponent = abs(other.m_value).numerator().convert_to<uint32_t>();
 			bigint numerator = pow(m_value.numerator(), exponent);
 			bigint denominator = pow(m_value.denominator(), exponent);
+
+			// Limit size to 4096 bits
+			if (numerator >= pow(bigint(2), rationalMaxBits) - 1)
```

Comment:
This `pow` should be evaluated in the const variable. Also you can use shift instead of pow for power-of-2.

Reply:
I don't think it is enough to check for the value **after** it has ben evaluated. It should be guessed before evaluating with `pow`.

One of the main problems here is the DoS vector (which is just annoying for a user, for example running the compiler in the browser).

Reply:
I don't think it is enough to check for the value **after** it has ben evaluated. It should be guessed before evaluating with `pow`.

One of the main problems here is the DoS vector (which is just annoying for a user, for example running the compiler in the browser).

Reply:
I've started testing with two examples of those DoS cases and they're not hanging anymore. After the limit is reached, the function is still called, but  ``m_value.denominator()`` is constant from there on and calling ``pow()`` results in the same large number. 
Currently the error messages don't help to understand the issue I guess, but there might be some options for makes those richer.